File: shared/timeline/global_settings.py
from __future__ import annotations

from copy import deepcopy
import json
import os
from pathlib import Path
from typing import Any, Mapping

import folder_paths

from ..contracts.video_timeline import (
    DEFAULT_ALLOW_GAPS,
    DEFAULT_AUTO_CLOSE_GAPS,
    DEFAULT_MINIMUM_SECTION_DURATION_SECONDS,
    DEFAULT_PROJECT_ASSET_ROOT_SUBDIRECTORY,
)


CONFIG_DIR = Path(__file__).resolve().parents[2] / "config"
SETTINGS_FILE_NAME = "timeline_director_global_settings.json"
GLOBAL_SETTINGS_SCHEMA_VERSION = 1


DEFAULT_GLOBAL_SETTINGS: dict[str, Any] = {
    "schema_version": GLOBAL_SETTINGS_SCHEMA_VERSION,
    "storage": {
        "asset_root_directory": "",
    },
    "timeline": {
        "show_resolved_model_output": False,
        "allow_gaps": DEFAULT_ALLOW_GAPS,
        "auto_close_gaps": DEFAULT_AUTO_CLOSE_GAPS,
        "minimum_section_duration_seconds": DEFAULT_MINIMUM_SECTION_DURATION_SECONDS,
    },
    "global_prompt": {
        "show_effective_prompt": False,
    },
    "audio": {
        "always_normalize": False,
    },
    "privacy": {
        "mode": True,
    },
    "display": {
        "show_section_labels": True,
        "show_thumbnails": True,
        "show_audio_waveforms": True,
    },
}


class GlobalSettingsError(ValueError):
    """Raised when Director global settings cannot be saved safely."""


def settings_path(base_dir: str | os.PathLike[str] | None = None) -> Path:
    root = Path(base_dir) if base_dir is not None else CONFIG_DIR
    return root / SETTINGS_FILE_NAME


def load_global_settings(base_dir: str | os.PathLike[str] | None = None) -> dict[str, Any]:
    path = settings_path(base_dir)
    if not path.exists():
        return default_global_settings()
    try:
        payload = json.loads(path.read_text(encoding="utf-8") or "{}")
    except Exception:
        payload = {}
    return normalize_global_settings(payload)


def save_global_settings(
    settings: Mapping[str, Any],
    base_dir: str | os.PathLike[str] | None = None,
) -> dict[str, Any]:
    normalized = normalize_global_settings(settings)
    configured_root = str(normalized["storage"]["asset_root_directory"]).strip()
    if configured_root and not Path(configured_root).expanduser().is_absolute():
        raise GlobalSettingsError("GLOBAL_ASSET_ROOT_NOT_ABSOLUTE: Asset Root Directory must be an absolute path.")
    path = settings_path(base_dir)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(normalized, indent=2, sort_keys=True), encoding="utf-8")
    return normalized


def patch_global_settings(
    patch: Mapping[str, Any],
    base_dir: str | os.PathLike[str] | None = None,
) -> dict[str, Any]:
    merged = _deep_merge(load_global_settings(base_dir), patch)
    return save_global_settings(merged, base_dir)


def global_settings_status(base_dir: str | os.PathLike[str] | None = None) -> dict[str, Any]:
    settings = load_global_settings(base_dir)
    default_root = default_asset_root_directory()
    effective_root = resolve_global_asset_root(settings=settings, create=False)
    return {
        "ok": True,
        "settings": settings,
        "configPath": str(settings_path(base_dir)),
        "storage": {
            "asset_root_directory": settings["storage"]["asset_root_directory"],
            "effective_asset_root_directory": str(effective_root),
            "default_asset_root_directory": str(default_root),
            "configured": bool(settings["storage"]["asset_root_directory"]),
        },
    }


def default_global_settings() -> dict[str, Any]:
    return deepcopy(DEFAULT_GLOBAL_SETTINGS)


def normalize_global_settings(value: Any) -> dict[str, Any]:
    source = value if isinstance(value, Mapping) else {}
    normalized = default_global_settings()
    normalized["schema_version"] = GLOBAL_SETTINGS_SCHEMA_VERSION
    normalized["storage"]["asset_root_directory"] = _safe_string(
        _section(source, "storage").get("asset_root_directory")
    )
    timeline = _section(source, "timeline")
    normalized["timeline"]["show_resolved_model_output"] = bool(timeline.get("show_resolved_model_output"))
    normalized["timeline"]["allow_gaps"] = timeline.get("allow_gaps") is not False
    normalized["timeline"]["auto_close_gaps"] = bool(timeline.get("auto_close_gaps"))
    normalized["timeline"]["minimum_section_duration_seconds"] = _positive_float(
        timeline.get("minimum_section_duration_seconds"),
        DEFAULT_MINIMUM_SECTION_DURATION_SECONDS,
    )
    normalized["global_prompt"]["show_effective_prompt"] = bool(
        _section(source, "global_prompt").get("show_effective_prompt")
    )
    normalized["audio"]["always_normalize"] = bool(_section(source, "audio").get("always_normalize"))
    normalized["privacy"]["mode"] = _section(source, "privacy").get("mode") is not False
    display = _section(source, "display")
    normalized["display"]["show_section_labels"] = display.get("show_section_labels") is not False
    normalized["display"]["show_thumbnails"] = display.get("show_thumbnails") is not False
    normalized["display"]["show_audio_waveforms"] = display.get("show_audio_waveforms") is not False
    return normalized


def default_asset_root_directory() -> Path:
    return (Path(folder_paths.get_output_directory()) / DEFAULT_PROJECT_ASSET_ROOT_SUBDIRECTORY).resolve()


def resolve_global_asset_root(
    *,
    settings: Mapping[str, Any] | None = None,
    create: bool = True,
) -> Path:
    normalized = normalize_global_settings(settings if settings is not None else load_global_settings())
    configured = _safe_string(normalized["storage"].get("asset_root_directory"))
    root = Path(configured).expanduser() if configured else default_asset_root_directory()
    if configured and not root.is_absolute():
        raise GlobalSettingsError("GLOBAL_ASSET_ROOT_NOT_ABSOLUTE: Asset Root Directory must be an absolute path.")
    root = root.resolve()
    if create:
        root.mkdir(parents=True, exist_ok=True)
    return root


def global_privacy_mode(settings: Mapping[str, Any] | None = None) -> bool:
    normalized = normalize_global_settings(settings if settings is not None else load_global_settings())
    return bool(normalized["privacy"]["mode"])


def global_always_normalize_audio(settings: Mapping[str, Any] | None = None) -> bool:
    normalized = normalize_global_settings(settings if settings is not None else load_global_settings())
    return bool(normalized["audio"]["always_normalize"])


def _section(source: Mapping[str, Any], key: str) -> Mapping[str, Any]:
    value = source.get(key)
    return value if isinstance(value, Mapping) else {}


def _deep_merge(base: Mapping[str, Any], patch: Mapping[str, Any]) -> dict[str, Any]:
    merged = deepcopy(dict(base))
    for key, value in patch.items():
        if isinstance(value, Mapping) and isinstance(merged.get(key), Mapping):
            merged[key] = _deep_merge(merged[key], value)
        else:
            merged[key] = deepcopy(value)
    return merged


def _positive_float(value: Any, fallback: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return float(fallback)
    if number <= 0:
        return float(fallback)
    return number


def _safe_string(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()

File: web/timeline/global_settings.js
import { ensureStoredPrivacyTokenCookie } from "./privacy.js";

export const GLOBAL_SETTINGS_ROUTE = "/helto_director/global_settings";

export const DEFAULT_GLOBAL_SETTINGS = {
  schema_version: 1,
  storage: {
    asset_root_directory: "",
    effective_asset_root_directory: "ComfyUI output/helto_director_projects",
    default_asset_root_directory: "ComfyUI output/helto_director_projects",
    configured: false,
  },
  timeline: {
    show_resolved_model_output: false,
    allow_gaps: true,
    auto_close_gaps: false,
    minimum_section_duration_seconds: 0.25,
  },
  global_prompt: {
    show_effective_prompt: false,
  },
  audio: {
    always_normalize: false,
  },
  privacy: {
    mode: true,
  },
  display: {
    show_section_labels: true,
    show_thumbnails: true,
    show_audio_waveforms: true,
  },
};

export function normalizeGlobalSettings(value = {}) {
  const source = value && typeof value === "object" && !Array.isArray(value) ? value : {};
  const payloadSettings = source.settings && typeof source.settings === "object" && !Array.isArray(source.settings)
    ? source.settings
    : source;
  const storageStatus = source.storage && typeof source.storage === "object" && !Array.isArray(source.storage)
    ? source.storage
    : {};
  const storage = section(payloadSettings, "storage");
  const timeline = section(payloadSettings, "timeline");
  const globalPrompt = section(payloadSettings, "global_prompt");
  const audio = section(payloadSettings, "audio");
  const privacy = section(payloadSettings, "privacy");
  const display = section(payloadSettings, "display");
  const assetRoot = stringValue(storage.asset_root_directory);
  const defaultRoot = stringValue(storageStatus.default_asset_root_directory) || DEFAULT_GLOBAL_SETTINGS.storage.default_asset_root_directory;
  const effectiveRoot = stringValue(storageStatus.effective_asset_root_directory) || assetRoot || defaultRoot;

  return {
    schema_version: 1,
    storage: {
      asset_root_directory: assetRoot,
      effective_asset_root_directory: effectiveRoot,
      default_asset_root_directory: defaultRoot,
      configured: Boolean(storageStatus.configured ?? assetRoot),
    },
    timeline: {
      show_resolved_model_output: Boolean(timeline.show_resolved_model_output),
      allow_gaps: timeline.allow_gaps !== false,
      auto_close_gaps: Boolean(timeline.auto_close_gaps),
      minimum_section_duration_seconds: positiveNumber(timeline.minimum_section_duration_seconds, DEFAULT_GLOBAL_SETTINGS.timeline.minimum_section_duration_seconds),
    },
    global_prompt: {
      show_effective_prompt: Boolean(globalPrompt.show_effective_prompt),
    },
    audio: {
      always_normalize: Boolean(audio.always_normalize),
    },
    privacy: {
      mode: privacy.mode !== false,
    },
    display: {
      show_section_labels: display.show_section_labels !== false,
      show_thumbnails: display.show_thumbnails !== false,
      show_audio_waveforms: display.show_audio_waveforms !== false,
    },
  };
}

export async function fetchGlobalSettings() {
  const response = await fetch(GLOBAL_SETTINGS_ROUTE, { cache: "no-store" });
  const payload = await response.json();
  if (!response.ok || payload?.ok === false || payload?.error) {
    throw new Error(payload?.error || "Failed to load global settings.");
  }
  return normalizeGlobalSettings(payload);
}

export async function saveGlobalSettings(settings) {
  ensureStoredPrivacyTokenCookie();
  const response = await fetch(GLOBAL_SETTINGS_ROUTE, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(settingsPayload(settings)),
  });
  const payload = await response.json();
  if (!response.ok || payload?.ok === false || payload?.error) {
    throw new Error(payload?.error || "Failed to save global settings.");
  }
  return normalizeGlobalSettings(payload);
}

export function settingsPayload(settings) {
  const normalized = normalizeGlobalSettings(settings);
  return {
    schema_version: normalized.schema_version,
    storage: {
      asset_root_directory: normalized.storage.asset_root_directory,
    },
    timeline: { ...normalized.timeline },
    global_prompt: { ...normalized.global_prompt },
    audio: { ...normalized.audio },
    privacy: { ...normalized.privacy },
    display: { ...normalized.display },
  };
}

export function isGlobalPrivacyMode(settings) {
  return normalizeGlobalSettings(settings).privacy.mode;
}

export function globalAssetRootLabel(settings) {
  const normalized = normalizeGlobalSettings(settings);
  return normalized.storage.effective_asset_root_directory || normalized.storage.default_asset_root_directory;
}

function section(source, key) {
  const value = source[key];
  return value && typeof value === "object" && !Array.isArray(value) ? value : {};
}

function stringValue(value) {
  return value == null ? "" : String(value).trim();
}

function positiveNumber(value, fallback) {
  const number = Number(value);
  return Number.isFinite(number) && number > 0 ? number : fallback;
}
